ci: add a backward compatibility checks to the end-to-end runner (#2816)
input-output-hk/mithril: mithril-test-lab/mithril-end-to-end/backward-compatibility.md

# Backward compatibility

The end-to-end tests scenarios are backward compatible with previous nodes versions.

This is achieved by differentiating the scenarios based on the version of the nodes they are testing.

There are multiple ways to do this:

- disabling part of the scenarios

> - adding or removing arguments to the binaries (i.e. `--backend` flag for the client-cli)

Reviewer: ```suggestion
- adding or removing arguments to the binaries (e.g. `--backend` flag for the client-cli)
```

> [!TIP]
> Node versions and distribution versions below are the ones that first included the changes, has node versions evolve
> faster than distribution versions the actual node version included in the distribution may be higher.
>
> i.e. `--origin-tag` flag for the client-cli was introduced in version `0.11.13`, the first distribution that included
> that change is `2517.0`, which include client-cli `0.12.0`.

---

## Supported changes

List of breaking changes that are supported by the end-to-end tests.

format is: `- since 'X.Y.Z' (distribution version) [to 'X.Y.Z' (distribution version) (optional)]: supported change`

### Mithril client

> - since `0.12.34` (2543.0): test of new `--epoch` filter to `cardano-db list` (disabled on lower versions)

Reviewer: This will make the lists a bit more readable:
```suggestion
- **since `0.12.34` (2543.0)**: test of new `--epoch` filter to `cardano-db list` (disabled on lower versions)
```

- since `0.12.11` (2524.0): removal of `cardano-db-v2` replaced with `cardano-db [command] --backend [v1,v2]`
- since `0.11.14` (2517.0): addition of `--include-ancillary` flag to `cardano-db download`
- since `0.11.13` (2517.0): addition of global `--origin-tag` parameter

### Mithril aggregator

- since `0.7.94` (next to 2543.1): only the leader aggregator must be restarted when updating protocol parameters

### Mithril signer

---

> ## Not supported changes

Reviewer: ```suggestion
## Unsupported changes
```

List of breaking changes that are NOT supported by the end-to-end tests, running a node with those versions against

> `unstable` will result in an error and an exit code of `3`.

Author: ```suggestion
List of breaking changes that are NOT supported by the end-to-end tests, running the tests with nodes versions
violating at least one of the case below will result in an error and an exit code of `3`.
```
